src/solvers/LevenbergMarquardt.jl

```diff
@@ -0,0 +1,166 @@
+    > and Local Convergence Properties.” arXiv, Oct. 01, 2022.
+    > doi: [10.48550/arXiv.2210.00253](https://doi.org/10.48550/arXiv.2210.00253).
+[^Peeters1993]:
+    > R. L. M. Peeters, “On a Riemannian version of the Levenberg-Marquardt algorithm,”
+    > VU University Amsterdam, Faculty of Economics, Business Administration and Econometrics,
+    > Serie Research Memoranda 0011, 1993.
+    > link: [https://econpapers.repec.org/paper/vuawpaper/1993-11.htm](https://econpapers.repec.org/paper/vuawpaper/1993-11.htm).
+"""
+function LevenbergMarquardt(
+    M::AbstractManifold, F::TF, gradF::TDF, x, num_components::Int; kwargs...
+) where {TF,TDF}
+    x_res = copy(M, x)
+    return LevenbergMarquardt!(M, F, gradF, x_res, num_components; kwargs...)
+end
+
+@doc raw"""
+    LevenbergMarquardt!(M, F, jacF, x)
+
+
+For more options see [`LevenbergMarquardt`](@ref).
+"""
+function LevenbergMarquardt!(
+    M::AbstractManifold,
+    F::TF,
+    jacF::TDF,
+    x,
+    num_components::Int;
+    retraction_method::AbstractRetractionMethod=default_retraction_method(M),
+    stopping_criterion::StoppingCriterion=StopAfterIteration(200) |
+                                          StopWhenGradientNormLess(1e-12) |
+                                          StopWhenStepsizeLess(1e-12),
+    debug=[DebugWarnIfCostIncreases()],
+    evaluation::AbstractEvaluationType=AllocatingEvaluation(),
+    flagnz::Bool=true,
+    jacB::AbstractBasis=DefaultOrthonormalBasis(),
+    kwargs..., #collect rest
+) where {TF,TDF}
+    p = NonlinearLeastSquaresProblem(
+        M, F, jacF, num_components; evaluation=evaluation, jacB=jacB
+    )
+    o = LevenbergMarquardtOptions(
+        M,
+        x,
+        similar(x, num_components),
+        similar(x, num_components, manifold_dimension(M));
+        stopping_criterion=stopping_criterion,
+        retraction_method=retraction_method,
+        flagnz=flagnz,
+    )
+    o = decorate_options(o; debug=debug, kwargs...)
+    return get_solver_return(solve(p, o))
+end
+#
+# Solver functions
+#
+function initialize_solver!(
+    p::NonlinearLeastSquaresProblem{AllocatingEvaluation}, o::LevenbergMarquardtOptions
+)
+    o.Fval = p.F(p.M, o.x)
+    o.gradient = get_gradient(p, o.x)
+    return o
+end
+function initialize_solver!(
+    p::NonlinearLeastSquaresProblem{MutatingEvaluation}, o::LevenbergMarquardtOptions
+)
+    p.F(p.M, o.Fval, o.x)
+    o.gradient = get_gradient(p, o.x)
+    return o
+end
+
+function _maybe_get_basis(M::AbstractManifold, p, B::AbstractBasis)
+    if requires_caching(B)
+        return get_basis(M, p, B)
+    else
+        return B
+    end
+end
+
+function step_solver!(
+    p::NonlinearLeastSquaresProblem{Teval}, o::LevenbergMarquardtOptions, iter::Integer
+) where {Teval<:AbstractEvaluationType}
+    # o.Fval is either initialized by initialize_solver! or taken from the previous iteraion
+
+    basis_ox = _maybe_get_basis(p.M, o.x, p.jacB)
+    if Teval === AllocatingEvaluation
```

Comment:
Good question, though here the meaning of Jacobian is tied to the nonlinear least squares problem so it doesn't seem as universal as `get_gradient`.

I will add documentation about `B_dom` somewhere.

Reply:
I've chosen `B_dom` because in general to calculate a Jacobian we need bases for both domain and codomain, though here I've opted to assume that in the codomain the standard Euclidean basis is used. It's more about potential consistency with cases where more general Jacobians might be useful. Maybe it's not worth it though.

Regarding the other point, I'm not sure how `get_jacobian` would help simplify `step_solver!`?

Reply:
`get_jacobian` would do the `if Teval === AllocatingEvaluation` via dispatch (as do all other functions accessing “stuff” from the problem).

Concerning the `B_dom` I would then maybe prefer `basis_domain`?

Reply:
OK, I will change the name to `basis_domain`.

Regarding `get_jacobian`, I think we have the same issue as in https://github.com/JuliaManifolds/Manopt.jl/pull/167#discussion_r1038007228 .

Reply:
Hm, again, maybe just “steal” from `get_gradient`?

Reply:
This is also resolved with the new variant.